sled-agent/src/instance.rs

```diff
@@ -516,37 +518,80 @@ impl Instance {
         info!(inner.log, "Created address {} for zone: {}", network, zname);
 
         // Run Propolis in the Zone.
+        let smf_service_name = "svc:/system/illumos/propolis-server";
+        let instance_name = format!("vm-{}", inner.propolis_id());
+        let smf_instance_name =
+            format!("{}:{}", smf_service_name, instance_name);
         let server_addr = SocketAddr::new(inner.propolis_ip, PROPOLIS_PORT);
+
+        // We intentionally do not import the service - it is placed under
+        // `/var/svc/manifest`, and should automatically be imported by
+        // configd.
+        //
+        // Insteady, we re-try adding the instance until it succeeds.
+        // This implies that the service was added successfully.
+        info!(
+            inner.log,
+            "Adding {} as a {} service", &instance_name, smf_service_name
+        );
+        backoff::retry_notify(
+            backoff::internal_service_policy(),
+            || async {
+                running_zone
+                    .run_cmd(&[
+                        crate::illumos::zone::SVCCFG,
+                        "-s",
+                        smf_service_name,
+                        "add",
+                        &instance_name,
+                    ])
+                    .map_err(|e| backoff::BackoffError::transient(e))
+            },
+            |err: RunCommandError, delay| {
+                warn!(
+                    inner.log,
+                    "Failed to add {} as a service (retrying in {:?}): {}",
+                    instance_name,
+                    delay,
+                    err.to_string()
+                );
+            },
+        )
+        .await?;
+
+        info!(inner.log, "Adding service property group");
         running_zone.run_cmd(&[
             crate::illumos::zone::SVCCFG,
-            "import",
-            "/var/svc/manifest/site/propolis-server/manifest.xml",
+            "-s",
+            &smf_instance_name,
+            "addpg",
+            "config",
+            "astring",
         ])?;
 
+        info!(inner.log, "Setting server address property");
         running_zone.run_cmd(&[
             crate::illumos::zone::SVCCFG,
             "-s",
-            "system/illumos/propolis-server",
+            &smf_instance_name,
             "setprop",
             &format!("config/server_addr={}", server_addr),
         ])?;
 
+        info!(inner.log, "Refreshing instance");
         running_zone.run_cmd(&[
             crate::illumos::zone::SVCCFG,
             "-s",
-            "svc:/system/illumos/propolis-server",
-            "add",
-            &format!("vm-{}", inner.propolis_id()),
+            &smf_instance_name,
+            "refresh",
```

Comment:
We had talked about putting an additional check here, that the `config/server_addr` property has actually taken for the instance. Do you think that's overly paranoid?

Reply:
I think I'm okay omitting it, based on the data collected (500+ iterations without the server_addr missing), but I'd be happy to add it later if we see any issues?